File: algorithms/permutations/number of steps/num_steps.py
""" 
You are climbing a staircase. It takes n steps to reach the top.
Each time, you can either climb 1 or 2 steps. In how many distinct ways can you climb to the top?
Note: Given n will be a positive integer.
"""
import time

def num_steps(current, destination):
    if current == destination:
        return 1
    elif current > destination:
        return 0

    return num_steps(current + 1, destination) + num_steps(current + 2, destination)

start = time.time()
permutations = num_steps(0, 20)
stop = time.time()
duration = stop - start
print(
    f"There are {permutations} permutations of 1 and/or 2 steps to reach destination. Computed in {duration} seconds")



def num_steps_memoized(current, destination, cache={}):
    if current == destination:
        return 1
    elif current > destination:
        return 0
    elif current not in cache:
        cache[current] = num_steps_memoized(current + 1, destination, cache) + num_steps_memoized(current + 2, destination, cache)

    return cache[current]

start2 = time.time()
permutations2 = num_steps_memoized(0, 200) # note input is 10x larger!
stop2 = time.time()
duration2 = stop2 - start2
print(
    f"There are {permutations2} permutations of 1 and/or 2 steps to reach destination. Computed in {duration2} seconds")

File: algorithms/permutations/number of steps/numSteps.js
function numSteps(current, destination) {
  if (current === destination) {
    return 1;
  } else if (current > destination) {
    return 0;
  }

  return (
    numSteps(current + 1, destination) + numSteps(current + 2, destination)
  );
}

const start = Date.now();
const result = numSteps(0, 20);
const stop = Date.now();
const duration = (stop - start) / 1000;
console.log(
  `There are ${result} permutations of 1 and/or 2 steps to reach destination. Computed in ${duration} seconds`
);

function numStepsMemoized(current, destination, cache = {}) {
  if (current === destination) {
    return 1;
  } else if (current > destination) {
    return 0;
  } else if (cache[current] === undefined) {
    cache[current] =
      numStepsMemoized(current + 1, destination, cache) +
      numStepsMemoized(current + 2, destination, cache);
  }
  return cache[current];
}

const start2 = Date.now();
const result2 = numStepsMemoized(0, 200); // note input is 10x larger!
const stop2 = Date.now();
const duration2 = (stop - start) / 1000;
console.log(
  `There are ${result2} permutations of 1 and/or 2 steps to reach destination. Computed in ${duration2} seconds`
);
